src/contracts/permissions/PermissionController.sol

```diff
@@ -0,0 +1,176 @@
+        require(admins.remove(admin), AdminNotSet());
+
+        emit AdminRemoved(account, admin);
+    }
+
+    /// @inheritdoc IPermissionController
+    function setDelegate(
+        address account,
+        address delegate,
+        address target,
+        bytes4 selector
+    ) external onlyAdmin(account) {
+        AccountPermissions storage permissions = _permissions[account];
+
+        bytes32 targetSelector = _encodeTargetSelector(target, selector);
+        require(!permissions.delegatePermissions[delegate].contains(targetSelector), DelegateAlreadySet());
+
+        // Add the delegate to the account's permissions
+        permissions.delegatePermissions[delegate].add(targetSelector);
+        permissions.permissionDelegates[targetSelector].add(delegate);
+
+        emit DelegateSet(account, delegate, target, selector);
+    }
+
+    /// @inheritdoc IPermissionController
+    function removeDelegate(
+        address account,
+        address delegate,
+        address target,
+        bytes4 selector
+    ) external onlyAdmin(account) {
+        AccountPermissions storage permissions = _permissions[account];
+
+        bytes32 targetSelector = _encodeTargetSelector(target, selector);
+        require(permissions.delegatePermissions[delegate].contains(targetSelector), DelegateNotSet());
+
+        // Remove the delegate from the account's permissions
+        permissions.delegatePermissions[delegate].remove(targetSelector);
+        permissions.permissionDelegates[targetSelector].remove(delegate);
+
+        emit DelegateRemoved(account, delegate, target, selector);
+    }
+
+    /**
+     *
+     *                         INTERNAL FUNCTIONS
+     *
+     */
+
+    /// @dev Encodes the target and selector into a single bytes32 value
+    function _encodeTargetSelector(address target, bytes4 selector) internal pure returns (bytes32) {
+        return bytes32(abi.encodePacked(target, uint96(bytes12((selector)))));
+    }
+
+    /// @dev Decodes the target and selector from a single bytes32 value
+    function _decodeTargetSelector(
+        bytes32 targetSelector
+    ) internal view returns (address, bytes4) {
+        address target = address(uint160(uint256(targetSelector) >> 96));
+        // The selector is in the lower 32 bits of the targetSelector
+        bytes4 selector = bytes4(uint32(uint256(targetSelector) >> 64));
+
+        return (target, selector);
+    }
+
+    /**
+     *
+     *                         VIEW FUNCTIONS
+     *
+     */
+
+    /// @inheritdoc IPermissionController
+    function isAdmin(address account, address caller) public view returns (bool) {
+        if (_permissions[account].admins.length() == 0) {
+            // If the account does not have an admin, the caller must be the account
+            return account == caller;
+        } else {
+            // If the account has an admin, the caller must be an admin
+            return _permissions[account].admins.contains(caller);
```

Comment:
this is done in constant lookup time so probably can just remove the length == 0 check. Remove the extra storage check for users who have admins set